injector/injector.go

```diff
@@ -0,0 +1,312 @@
+	wrapperTagFound := false
+	impressionTagFound := false
+	errorTagFound := false
+	creativeId := ""
+	isCreative := false
+	companionTagFound := false
+	nonLineaerTagFound := false
+
+	b := strings.NewReader(vastXML)
+	p := xml.NewDecoder(b)
+
+	for {
+		t, err := p.RawToken()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return ""
+		}
+
+		switch tt := t.(type) {
+		case xml.StartElement:
+			switch tt.Name.Local {
+			case "Wrapper":
+				wrapperTagFound = true
+			case "Creative":
+				isCreative = true
+				for _, attr := range tt.Attr {
+					if strings.ToLower(attr.Name.Local) == "adid" {
+						creativeId = attr.Value
+					}
+				}
+			case "Linear":
+				injectVideoClicks = true
+				injectTracker = true
+			case "VideoClicks":
+				injectVideoClicks = false
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				ti.provider.PopulateEventMacros(creativeId, "", "")
+
+				for _, url := range ti.events.VideoClicks {
+					outputXML.WriteString("<ClickTracking><![CDATA[")
+					ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+					outputXML.WriteString("]]></ClickTracking>")
+				}
+
+				continue
+			case "NonLinearAds":
+				injectTracker = true
+			case "TrackingEvents":
+				if isCreative {
+					injectTracker = false
+					encoder.Flush()
+					encoder.EncodeToken(tt)
+					encoder.Flush()
+					for typ, urls := range ti.events.LinearTrackingEvents {
+						ti.provider.PopulateEventMacros(creativeId, "tracking", typ)
+						for _, url := range urls {
+							outputXML.WriteString("<Tracking event=\"")
+							outputXML.WriteString(string(typ))
+							outputXML.WriteString("\"><![CDATA[")
+							ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+							outputXML.WriteString("]]></Tracking>")
+						}
+					}
+					continue
+				}
+			}
+
+		case xml.EndElement:
+			switch tt.Name.Local {
+			case "Impression":
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				fmt.Println()
+				if !impressionTagFound {
+					for _, url := range ti.events.Impressions {
+						outputXML.WriteString("<Impression><![CDATA[")
+						ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+						outputXML.WriteString("]]></Impression>")
+					}
+					impressionTagFound = true
+				}
+				continue
+			case "Error":
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				if !errorTagFound {
+					for _, url := range ti.events.Errors {
+						outputXML.WriteString("<Error><![CDATA[")
+						ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+						outputXML.WriteString("]]></Error>")
+
+					}
+					errorTagFound = true
+				}
+				continue
+			case "NonLinearAds":
+				if injectTracker {
+					injectTracker = false
+					encoder.Flush()
+					outputXML.WriteString("<TrackingEvents>")
+					for typ, urls := range ti.events.LinearTrackingEvents {
+						ti.provider.PopulateEventMacros(creativeId, "", typ)
+						for _, url := range urls {
+
```

Comment:
Nitpick: Can you remove this blank line?

Reply:
Done